pkg/common/flyte_url.go

```diff
@@ -0,0 +1,98 @@
+package common
+
+import (
+	"fmt"
+	"regexp"
+	"strconv"
+
+	"github.com/flyteorg/flyteidl/gen/pb-go/flyteidl/admin"
+	"github.com/flyteorg/flyteidl/gen/pb-go/flyteidl/core"
+)
+
+// transform to snake case to make lower case
+//go:generate enumer --type=ArtifactType --trimprefix=ArtifactType -transform=snake
+
+type ArtifactType int
+
+// The suffixes in these constants are used to match against the tail end of the flyte url, to keep tne flyte url simpler
+const (
+	ArtifactTypeUndefined ArtifactType = iota
+	ArtifactTypeI                      // inputs
+	ArtifactTypeO                      // outputs
+	ArtifactTypeD                      // deck
+)
+
+var re = regexp.MustCompile("flyte://v1/(?P<project>[a-zA-Z0-9_-]+)/(?P<domain>[a-zA-Z0-9_-]+)/(?P<exec>[a-zA-Z0-9_-]+)/(?P<node>[a-zA-Z0-9_-]+)(?:/(?P<attempt>[0-9]+))?/(?P<artifactType>[iod])$")
+
+func MatchRegex(reg *regexp.Regexp, input string) map[string]string {
+	names := reg.SubexpNames()
+	res := reg.FindAllStringSubmatch(input, -1)
+	if len(res) == 0 {
+		return nil
+	}
+	dict := make(map[string]string, len(names))
+	for i := 1; i < len(res[0]); i++ {
+		dict[names[i]] = res[0][i]
+	}
+	return dict
+}
+
+func ParseFlyteURL(flyteURL string) (core.NodeExecutionIdentifier, *int, ArtifactType, error) {
+	// flyteURL is of the form flyte://v1/project/domain/execution_id/node_id/attempt/[iod]
+	// where i stands for inputs.pb o for outputs.pb and d for the flyte deck
+	// If the retry attempt is missing, the io requested is assumed to be for the node instead of the task execution
+	matches := MatchRegex(re, flyteURL)
+	proj := matches["project"]
+	domain := matches["domain"]
+	executionID := matches["exec"]
+	nodeID := matches["node"]
+	var attempt *int // nil means node execution, not a task execution
+	if matches["attempt"] != "" {
+		a, err := strconv.Atoi(matches["attempt"])
+		if err != nil {
+			return core.NodeExecutionIdentifier{}, nil, ArtifactTypeUndefined, fmt.Errorf("failed to parse attempt, %s", err)
+		}
```

Comment:
nit, also print the value of attempt
```suggestion
	if attempt := matches["attempt"]; len(attempt) > 0 {
		a, err := strconv.Atoi(attempt)
		if err != nil {
			return core.NodeExecutionIdentifier{}, nil, ArtifactTypeUndefined, fmt.Errorf("failed to parse attempt [%v], %v", attempt, err)
		}
```